### ISSUE TITLE
Deprecate () to mount in favor of []

### ISSUE THREAD

**Issue by catmando** [ruby-hyperloop/hyper-react#234]
I guess before raising this issue I should try this first :-)

But I am thinking that saying`DIV[class: :foo]` is much cheaper to implement than `DIV(class: :foo)`...

but I am not 100% sure about it... seems like something @janbiedermann would like to look at :-)

If memory serves the way `DIV(...)` is implemented is a bit bizarre, and I can only explain with an example:

class MyComponent < React::Component
  render do
    DIV(...)
  end
end

What goes on here is that DIV needs to act like an instance_method on React::Component.  So when DIV or any other component is defined, that method is added to React::Component (which everybody inherits from).   But it gets more complicated if the component is defined in a module.  In this case all component should be defined as a method in module!  

LIke I said I could be wrong about this, as some of the code has gotten simplified and cleaned up, but bottom line simply adding a class method `[]` would involve much less meta programming, and I am guessing would be faster, etc.

I think its also just less strange.  Its pretty standard ruby to say Foo[...], but you never actually see Foo(...)

So I am proposing deprecating `()` right now, so that we can get rid of it asap.

**Comment by catmando** [comment]
By the way I know that this was much harder to implement effeciently for Operations (because MyOperation() ) has to be defined for any object.   So although I know everybody loves saying: MyOp.run(...) we could actually shorten this to MyOp[...] which I personally really like :-)

**Comment by janbiedermann** [comment]
For operations:
MyOp.run() is it now, and i have MyOp.run_sync() on my list.
With MyOp[] this could then become: MyOp[sync:true]?

For Components:
The tags get defined as instance methods, this code is actually executed only once and then optimised away by the js engine, so its very fast. Then they are available as instance methods, so calling them is fast too.
With [] tags would need to be classes. Probably thats ok for performance too, and gives us other benefits, as the tags now can have a render method or a render_to_string or hydrate or whatever else react comes up with, even a should_component_update maybe, like one of my ideas to make tags components too, native ones even.

Components like MyComponent() are resolved by method_missing. This is can be slow.
MyComponent[] would be much better here for the overall performance of the client.

Native JS components are called the same way? Now const_missing would have to create a wrapper with define_method '[]' , or how do you imagine calling native components?

**Comment by catmando** [comment]
`MyOp[sync: true]` is not a good idea for syntax, because then the programmer can't use sync as an argument to the operation.

Remind me what does sync do?  Depending on what it does you could do `MyOp.sync(...)` or possibly `MyOp[...].sync`.   (I do wish ruby had a few more operators to play with like C++)

Right so built in tags, makes no difference by application components should be better performance.

Native components would be exactly as you say.  I can't remember the details of who we find native component names, but it would just be modified to work with `[]`  .  No matter what if we made it work with `()` we can certainly make it work (and probably faster) with `[]`.

**Comment by janbiedermann** [comment]
sync makes the caller wait for the operation to finish, synchronously, as needed for the console.

so we could have then MyOp[], equal to MyOp.run(), and as alternative MyOp.run_sync() or MyOp.sync().

**Comment by catmando** [comment]
How does it make caller wait?  Using synchronous http request s?   If so perhaps should just be a subclass of serverop?

**Comment by janbiedermann** [comment]
yes, sync http reqs, maybe a subclass, but i think wether the Op should be sync or async is a requirement of the caller not the operation, the operation doesn't care and should not care how its called, but the caller may have such requirement, depending on a condition even (like prerender or not?). So it should be determinable by the caller, i think.
And then the server should decide if it actually fulfils the request in sync or not.

**Comment by janbiedermann** [comment]
Looking into sync Operations further, it seems, to make this reliably work "WebWorkers" need to be introduced, as in recent Standard updates, sync requests are only allowed within WebWorkers. Using WebWorkers maybe in general a nice Idea for Operations on the client, maybe for 2.0, currently too much effort.

For () -> [] maybe something like a "upgrade helper" that converts component may be useful.
Rails hyperloop:assist_upgrade or something.

**Comment by catmando** [comment]
I am very sad about this Sync request... I was hoping to use it to make a real hyper-console that could stop execution!

**Comment by catmando** [comment]
[] doesn't work cause it can't take a block :-(
